contracts/cw4973/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+
+    // verify the signature using the hash and the public key
+    let is_verified = deps.api.secp256k1_verify(&hash, &sig, &pubkey_bytes);
+
+    match is_verified {
+        Ok(true) => {
+            // If the signature is verified then we must check the address of public key is equal to passive address
+            // get hrp from the signature
+            let hrp = signature.hrp.clone();
+
+            // get the address of signer from the public key using get_bech32_address function
+            let signer_address = _get_bech32_address(&hrp, pubkey_bytes.as_slice()).unwrap();
+
+            if signer_address != *passive {
+                Err(ContractError::InvalidSigner)
+            } else {
+                // return hex encoded hash
+                Ok(hex::encode(hash))
+            }
+        }
+        Ok(false) => Err(ContractError::InvalidSignature),
+        Err(_) => Err(ContractError::CannotVerifySignature),
+    }
+}
+
+// the get_hash funtion will concat the address of the sender, the address of the 'to', the uri of the nft and the hash of the string
+fn _get_hash(active: &str, passive: &str, uri: &str, chain_id: &str) -> Vec<u8> {
+    // hash the constant string and data
+    let big_string = base64::encode(format!(
+        "{}{}{}{}{}",
+        AGREEMENT_STRING, chain_id, active, passive, uri
+    ));
+
+    // get the signing document
+    let sign_doc_json = _get_sign_doc(passive, &big_string);
+
+    let hash = Sha256::digest(sign_doc_json.as_bytes());
+
+    // return the hash
+    return hash.as_slice().to_vec();
+}
+
+// rewrite mint function of cw721 base to ignore minter checking
+fn _mint(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: MintMsg<Extension>,
+) -> Result<Response, ContractError> {
+    // create the token
+    let token = TokenInfo {
+        owner: deps.api.addr_validate(&msg.owner)?,
+        approvals: vec![],
+        token_uri: msg.token_uri,
+        extension: msg.extension,
+    };
+
+    // update tokens list of contract
+    Cw4973Contract::default()
+        .tokens
+        .update(deps.storage, &msg.token_id, |old| match old {
+            Some(_) => Err(ContractError::Claimed),
+            None => Ok(token),
+        })?;
+    Cw4973Contract::default().increment_tokens(deps.storage)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "mint")
+        .add_attribute("minter", info.sender)
+        .add_attribute("owner", msg.owner)
+        .add_attribute("token_id", msg.token_id))
+}
+// OR: create new contract to handle the signing document is better
+// create signable structure from message and chain ID
+// @param signer: the address of the signer
+// @param message: the message to sign
+// @param chain_id: the chain id of the chain
+// @return: the signable structure
+// TODO: modify this function to specify the others fields of the signing document
+fn _get_sign_doc(signer: &str, message: &str) -> String {
```

Comment:
this and your types in state.rs are 🙏.

1. is it possible to write an integration test for this? using cosmjs from the readme would be great.
2. if you'd like, i think this signing logic would be a very helpful package to have. would you be interested in breaking it out into one?